storage/olddbclean/oldDatabaseCleaner.go

```diff
@@ -0,0 +1,231 @@
+		log.Debug("cannot delete old databases as the previous epoch does not exist in configuration",
+			"epoch", epochForDeletion,
+			"epochs configuration", odc.oldestEpochsToKeep)
+		return false
+	}
+	if previousOldestEpoch > newOldestEpoch {
+		log.Debug("skipping cleaning of old databases because new oldest epoch is older than previous",
+			"previous older epoch", previousOldestEpoch,
+			"actual older epoch", newOldestEpoch)
+		return false
+	}
+
+	return true
+}
+
+func (odc *oldDatabaseCleaner) computeOldestEpochToKeep(currentEpoch uint32) uint32 {
+	odc.Lock()
+	defer odc.Unlock()
+
+	oldestEpoch := uint32(math.MaxUint32)
+	storers := odc.storageListProvider.GetAllStorers()
+	for _, storer := range storers {
+		localEpoch, err := storer.GetOldestEpoch()
+		if err != nil {
+			continue
+		}
+
+		if localEpoch < oldestEpoch {
+			oldestEpoch = localEpoch
+		}
+	}
+
+	if oldestEpoch == uint32(math.MaxUint32) {
+		oldestEpoch = currentEpoch
+	}
+
+	return oldestEpoch
+}
+
+func (odc *oldDatabaseCleaner) cleanOldEpochs(currentEpoch uint32) error {
+	odc.Lock()
+	defer odc.Unlock()
+
+	epochForDeletion := currentEpoch - 1
+	epochToDeleteTo := odc.oldestEpochsToKeep[epochForDeletion]
+
+	epochDirectories, err := odc.directoryReader.ListDirectoriesAsString(odc.databasePath)
+	if err != nil {
+		return err
+	}
+
+	if len(epochDirectories) == 0 {
+		return nil
+	}
+
+	sortedEpochDirectories, sortedEpochs, found := getSortedEpochDirectories(epochDirectories)
+	if !found {
+		return nil
+	}
+
+	for idx, epoch := range sortedEpochs {
+		if epoch >= epochToDeleteTo {
+			break
+		}
+
+		fullDirectoryPath := path.Join(odc.databasePath, sortedEpochDirectories[idx])
+		log.Debug("removing old database", "db path", fullDirectoryPath)
+		err := odc.fileRemover(fullDirectoryPath)
+		if err != nil {
+			log.Warn("cannot remove old DB", "path", fullDirectoryPath, "error", err)
+		}
+	}
+
+	delete(odc.oldestEpochsToKeep, epochForDeletion)
+
+	return nil
+}
+
+func getSortedEpochDirectories(directories []string) ([]string, []uint32, bool) {
+	epochToDirMap := make(map[uint32]string)
+	epochs := make([]uint32, 0)
+	for _, dir := range directories {
+		epoch, ok := extractEpochFromDirName(dir)
+		if !ok {
+			continue
+		}
+
+		epochToDirMap[epoch] = dir
+		epochs = append(epochs, epoch)
+	}
+
+	sort.SliceStable(epochs, func(i, j int) bool {
+		return epochs[i] < epochs[j]
+	})
+
+	sortedDirectories := make([]string, 0, len(epochs))
+	for _, epoch := range epochs {
+		sortedDirectories = append(sortedDirectories, epochToDirMap[epoch])
+	}
+
+	notOkResults := len(sortedDirectories) != len(epochs) || len(sortedDirectories) == 0 || len(epochs) == 0
+	if notOkResults {
+		return nil, nil, false
+	}
+
+	return sortedDirectories, epochs, true
+}
+
+func extractEpochFromDirName(dirName string) (uint32, bool) {
+	splitStr := strings.Split(dirName, "_")
+	if len(splitStr) != 2 {
+		return 0, false
+	}
+
+	epoch, err := strconv.ParseInt(splitStr[1], 10, 64)
```

Comment:
magic numbers

Reply:
they are pretty straight-forward. we can think of creating a function in `core` that handles this as the same code is found in other 11 places